### DISCUSSION TITLE
Trackpad gestures

### DISCUSSION THREAD

**Discussion by ghost** [ghostty-org/ghostty#3506] [Feature Requests, Ideas]
It'd be really nice to see native gestures in Ghostty, something I feel lacking in most terminals, including the default Terminal.app. One out of many examples, pinching inwards should show all tabs - try this in Safari. I'm a GUI person, and I use the trackpad on my MacBook, so gestures is like a god sent for me - I use Swish which is a native gesture based window manager for macOS and it works flawlessly, but when the applications i use don't have gestures it feels like the flow of my productivity is cut in half.

**Comment by mitchellh** [comment]
As a very focused thing: pinching to see tab overview makes a ton of sense to me.

Are there other gestures that make sense to add? (That are idiomatic)

**Reply by ghost** [reply]
Gestures for navigating tabs comes to mind, think a two finger swipe (left and right) to navigate back and forth would be appropriate - in safari this would navigate back and forth the page history.

**Reply by dgnorton** [reply]
Another vote for 2 finger swipe to scroll left/right through panes.

**Comment by ghost** [comment]
I've never used iTerm but given it's also a native app it'd be interesting to hear what for, how and if they do gestures. so if anyone has experience with iterm and would like to share, please do! :)

**Reply by zachary822** [reply]
iTerm does three finger swipe left and right to go to prev and next tab. Three finger swipe up and down to cycle between windows.

I use the left/right three finger swipe quite often.

**Comment by jasondinh** [comment]
I'm currently using BetterTouchTool to map the gestures to their respective shortcuts, but it would be nice to get this supported without 3rd party tools.

**Comment by pluiedev** [comment]
From a technical perspective this would require a bit of extra code for GTK, since the builtin gesture event controllers support at most two-finger gestures (which is unfortunately due to legacy X11 support — X11 doesn't comprehend gestures involving more than two touch points). We would then therefore need to write our own event controller for more complex gestures.

**Comment by linardsblk** [comment]
Honestly just the two finger swipes for switching tabs would already be amazing. This is integrated into my workflow so much that it is the only nitpick stopping me from fully switching over.

**Reply by pluiedev** [reply]
Actually looking at GTK APIs a bit more, the only gestures that come out of the box are two-finger rotate, two-finger zoom and one-finger touches... That's certainly not going to be enough — we would most likely need to write our own gesture controller then

**Comment by pszypowicz** [comment]
I have wrote a PoC (MacOS) fork that I am daily driving. If someone want to try and give feedback on: 
https://github.com/pszypowicz/ghostty

Default gestures: 
| Gesture | Default action |
| --- | --- |
| `pinch-out` | `increase_font_size:1` |
| `pinch-in` | `decrease_font_size:1` |
| `smart-magnify` (two-finger double tap) | `toggle_font_zoom:6` |
| `swipe-left:2` | `next_tab` |
| `swipe-right:2` | `previous_tab` |

Discussion: https://github.com/ghostty-org/ghostty/discussions/13446
And Vouch Request: https://github.com/ghostty-org/ghostty/discussions/13446

This also allows to define gestures protocol that could be consumed by the cli apps:
https://github.com/tmux/tmux/issues/5462
https://github.com/pszypowicz/tmux

(and ofc: Claude Code, mainly Fable 5)
